sipi/scripts/upload.lua

```diff
@@ -42,111 +42,126 @@ if token == nil then
+    success, hashed_tmp_storage_filename = helper.filename_hash(tmp_storage_filename)
 
     if not success then
-        send_error(500, "Unable to create hashed filename: " .. tostring(hashed_jp2_filename))
+        send_error(500, "helper.filename_hash() failed: " .. tostring(hashed_tmp_storage_filename))
         return
     end
 
-    local jp2_file_path = config.imgroot .. '/tmp/' .. hashed_jp2_filename
-    local success, error_msg = uploaded_image:write(jp2_file_path)
+    local tmp_storage_file_path = config.imgroot .. '/tmp/' .. hashed_tmp_storage_filename
 
-    if not success then
-        send_error(500, "Unable to write " .. tostring(jp2_file_path) .. ": " .. tostring(error_msg))
-        return
-    end
+    -- Create a IIIF base URL for the converted file.
+    local tmp_storage_url = get_external_protocol() .. "://" .. get_external_hostname() .. ":" .. get_external_port() .. '/tmp/' .. tmp_storage_filename
 
-    server.log("upload.lua: wrote JPEG 2000 file to " .. jp2_file_path, server.loglevel.LOG_DEBUG)
+    -- Construct response data about the file that was uploaded.
 
+    local media_type = file_info["media_type"]
+    local this_file_upload_data = {}
+    this_file_upload_data["internalFilename"] = tmp_storage_filename
+    this_file_upload_data["originalFilename"] = original_filename
+    this_file_upload_data["temporaryUrl"] = tmp_storage_url
+    this_file_upload_data["fileType"] = media_type
+    file_upload_data[file_index] = this_file_upload_data
+
+    -- Is this an image file?
+    if media_type == IMAGE then
+        -- Yes. Create a new Lua image object. This reads the image into an
+        -- internal in-memory representation independent of the original
+        -- image format.
+        local uploaded_image
+        success, uploaded_image = SipiImage.new(file_index)
+
+        if not success then
+            send_error(500, "SipiImage.new() failed: " .. tostring(uploaded_image))
+            return
+        end
+
+        -- Check that the file extension is correct for the file's MIME type.
+        local check
+        success, check = uploaded_image:mimetype_consistency(mime_type, original_filename)
+
+        if not success then
+            send_error(500, "uploaded_image:mimetype_consistency() failed: " .. check)
+            return
+        end
+
+        if not check then
+            send_error(400, MIMETYPES_INCONSISTENCY)
+            return
+        end
+
+        -- Convert the image to JPEG 2000 format.
+
+        success, error_msg = uploaded_image:write(tmp_storage_file_path)
+
+        if not success then
+            send_error(500, "uploaded_image:write() failed for " .. tostring(tmp_storage_file_path) .. ": " .. tostring(error_msg))
+            return
+        end
+
+        server.log("upload.lua: wrote image file to " .. tmp_storage_file_path, server.loglevel.LOG_DEBUG)
+    else
+        -- It's not an image file. Just move it to its temporary storage location.
```

Comment:
Can Sipi check that the mime type is correct?

Reply:
I think this is only implemented for `SipiImage` in Sipi. Is there a more generic implementation needed for any file type? The current implementation relies on `magic.h`.

Reply:
Not as far as I know. We could open an issue for this in Sipi.

Reply:
Yes, please do so and assign it to me.

Reply:
#1530

Reply:
https://github.com/dasch-swiss/sipi/issues/316